### PR TITLE
SNAPSHOT: Improve Resilience SnapshotShardService

### PR DESCRIPTION
**Pull request by original-brownbear**
* Resolve the index in the snapshot thread
  * This is the only production code change in this PR. It's not a clean fix for #32265 but:
     * It works around the issue for now until the snapshot-resiliency feature branch gets merged
     * Resolving the `indexId` in the snapshot thread improves resilience in general because it covers any situations where the interaction with `indicesService` runs into an exception but that exception is never reported to the master
* Added test for routing table - snapshot state mismatch that reproduces #32265

### CODE REVIEW

**Comment by original-brownbear** [comment]
@ywelsch I realized this fixes the issue in #32265 without really changing behaviour in other situations.
What do you think about using this as a short-cut to fixing the issue while we figure out the best way to test this more granularly on the cluster-state level?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java [hunk 1]
```diff
@@ -323,15 +323,15 @@ private void processIndexShardSnapshots(ClusterChangedEvent event) {
...
-                    final IndexId indexId = indicesMap.get(shardId.getIndexName());
-                    assert indexId != null;
                     executor.execute(new AbstractRunnable() {
 
                         final SetOnce<Exception> failure = new SetOnce<>();
 
                         @Override
                         public void doRun() {
+                            final IndexShard indexShard = indicesService.indexServiceSafe(shardId.getIndex()).getShardOrNull(shardId.id());
+                            final IndexId indexId = indicesMap.get(shardId.getIndexName());
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java]: this can and should still be resolved outside the executor.

**Review by ywelsch — APPROVED** [review]
OK, let's do this (and also backport to 6.5).

**Comment by original-brownbear** [comment]
Jenkins test this

**Comment by original-brownbear** [comment]
Backported in #36166 and #36164 :)

**Pull request merged**